# Bibliotecas

In [1]:
import os
import pandas as pd
import numpy as np
import scipy as sp
from scipy import signal
import timeit
from datetime import datetime
from sklearn.externals import joblib
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
from sklearn import preprocessing
from sklearn import utils

from ml_time_series import generate_envelope

%matplotlib notebook

In [2]:
def HStacker(files_,t):
    
    #Lê dados do arquivo e remove janela
    Xc = np.load('DB_REPO/preproc/'+files_[3])
    Xc = Xc[:-49] #Exclui do STD para igualar o Hshape
    Xci = np.load('DB_REPO/preproc/'+files_[0])
    Xci = Xci[:-49]
    Xc = np.hstack((Xc.reshape(Xc.shape),Xci.reshape(Xci.shape)))
    
    #Aplica Hstack nos dados
    for j in range(1,len(files_)-1):
        Xci = np.load('DB_REPO/preproc/'+files_[j])
        Xc = np.hstack((Xc.reshape(Xc.shape),Xci.reshape(Xci.shape)))
          
    
    # Salving in file on the folder <DB_PKS/preproc/labeled>
    np.save('./DB_REPO/stacked/'+t+'_hs', Xc, allow_pickle=False)
    DATA_hs.append(t+'_hs.npy')

In [3]:
def VStacker(files_,t):
    
    #Lê dados do arquivo e remove janela
    Xc = np.load('DB_REPO/stacked/'+files_[0])
    
    #Aplica Hstack nos dados
    for j in range(0,len(files_)-1):
        Xci = np.load('DB_REPO/stacked/'+files_[j])
        Xc = np.vstack((Xc,Xci)) 
          
    
    # Salving in file on the folder <DB_PKS/preproc/labeled>
    np.save('./DB_REPO/stacked/'+t+'_vs', Xc, allow_pickle=False)
    DATA_vs.append(t+'_vs.npy')

In [4]:
def HStackerF(files_,t):
    
    #Lê dados do arquivo e remove janela
    Xc = np.load('DB_REPO/stacked/'+files_[5])
    Xc = Xc[:35853] #Exclui amostras para igualar tamanho e possibilitar o vstack
    
    #Aplica Hstack nos dados
    for j in range(0,len(files_)-1):
        Xci = np.load('DB_REPO/stacked/'+files_[j])
        Xci = Xci[:35853] #Exclui amostras para igualar tamanho e possibilitar o vstack
        Xc = np.hstack((Xc.reshape(Xc.shape),Xci.reshape(Xci.shape)))
          
    
    # Salving in file on the folder <DB_PKS/preproc/labeled>
    np.save('./DB_REPO/stacked/'+t+'_fs', Xc, allow_pickle=False)
    DATA_fs.append(t+'_fs.npy')

In [5]:
#Carregando arquivos
import commands

n_features = 4
A = commands.getoutput('ls DB_REPO/preproc')
files_ = A.split('\n')
DATA_hs = []

for i in range(0,len(files_),n_features):
    HStacker(files_[i:i+n_features],files_[i:i+n_features][0].split('-')[0]+str(i/n_features))
    #print (files_[i:i+3],files_[0:3][0].split('-')[0]+str(i/3))

In [6]:
(files_[i:i+n_features],files_[i:i+n_features][0].split('-')[0]+str(i/n_features))

(['PK_Sergio Lamenza - Dia 17.11.25 as 16.59.00.edf_fft.npy',
  'PK_Sergio Lamenza - Dia 17.11.25 as 16.59.00.edf_filt.npy',
  'PK_Sergio Lamenza - Dia 17.11.25 as 16.59.00.edf_rms.npy',
  'PK_Sergio Lamenza - Dia 17.11.25 as 16.59.00.edf_std.npy'],
 'PK_Sergio Lamenza 83')

In [7]:
files_[i:i+n_features][0].split('-')[0]+str(i/n_features)

'PK_Sergio Lamenza 83'

In [8]:
files_[0:][0].split('-')[0]

'CT_Antonia Perpetuo Borges '

In [9]:
files_[0:4]

['CT_Antonia Perpetuo Borges - Dia 17.12.11 as 08.56.09.edf_fft.npy',
 'CT_Antonia Perpetuo Borges - Dia 17.12.11 as 08.56.09.edf_filt.npy',
 'CT_Antonia Perpetuo Borges - Dia 17.12.11 as 08.56.09.edf_rms.npy',
 'CT_Antonia Perpetuo Borges - Dia 17.12.11 as 08.56.09.edf_std.npy']

In [10]:
for i in range(0,len(files_),4):
    r = files_[i:i+4]
    
    print (((np.load('DB_REPO/preproc/'+r[0])).shape[0],(np.load('DB_REPO/preproc/'+r[1])).shape[0],
            (np.load('DB_REPO/preproc/'+r[2])).shape[0],(np.load('DB_REPO/preproc/'+r[3])).shape[0]))

(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(14000, 13951, 13951, 14000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(22000, 21951, 21951, 22000)
(14000, 13951, 13951, 14000)
(12000, 11951, 11951, 12000)
(12000, 11951, 11951, 12000)
(22000, 21951,

In [11]:
import re 

print  (re.findall('\d*\D+',(DATA_hs[0:5][0]))[0])



CT_Antonia Perpetuo Borges 


In [12]:
DATA_hs

['CT_Antonia Perpetuo Borges 0_hs.npy',
 'CT_Antonia Perpetuo Borges 1_hs.npy',
 'CT_Antonia Perpetuo Borges 2_hs.npy',
 'CT_Francisco das Chagas dos Santos 3_hs.npy',
 'CT_Francisco das Chagas dos Santos 4_hs.npy',
 'CT_Francisco das Chagas dos Santos 5_hs.npy',
 'CT_Gelson Amantos de Souza  6_hs.npy',
 'CT_Gelson Amantos de Souza  7_hs.npy',
 'CT_Gelson Amantos de Souza  8_hs.npy',
 'CT_Geraldo Dias de Oliveira 9_hs.npy',
 'CT_Geraldo Dias de Oliveira 10_hs.npy',
 'CT_Geraldo Dias de Oliveira 11_hs.npy',
 'CT_Iraci Rodrigues Soares 12_hs.npy',
 'CT_Iraci Rodrigues Soares 13_hs.npy',
 'CT_Iraci Rodrigues Soares 14_hs.npy',
 'CT_Joaquina de Sousa Freitas  15_hs.npy',
 'CT_Joaquina de Sousa Freitas  16_hs.npy',
 'CT_Joaquina de Sousa Freitas  17_hs.npy',
 'CT_Jose Martins de Sousa  18_hs.npy',
 'CT_Jose Martins de Sousa  19_hs.npy',
 'CT_Jose Martins de Sousa  20_hs.npy',
 'CT_Larcir Iria da Conceicao Santos 21_hs.npy',
 'CT_Larcir Iria da Conceicao Santos 22_hs.npy',
 'CT_Larcir Iria d

In [13]:
Xc = np.load('DB_REPO/stacked/PK_Arthur Araujo Filho 42_hs.npy')

In [14]:
dfn = pd.DataFrame(data=Xc)
dfn.head()

0         1         2         3           4          5           6   \
0  0.022327 -0.001221  0.186603  0.017612  175.605680  10.642500  591.227547   
1  0.019318 -0.001221  0.186603  0.015084   75.150074   4.178106  409.107829   
2  0.025335  0.000705  0.173561  0.024354  120.955158   1.798792  822.637772   
3  0.016310 -0.000258  0.173561  0.013398   76.398090   2.421141  242.650976   
4  0.022327  0.000224  0.199644  0.015926   46.518462   1.619408  259.992532   

          7         8         9         10        11        12        13  \
0  50.701473  0.000375  0.000054 -0.002637  0.000031  0.024166  0.004446   
1  23.394437  0.000290  0.000073 -0.001959 -0.000076  0.024076  0.004444   
2  21.162750  0.000312  0.000108 -0.001131 -0.000242  0.023946  0.004452   
3  33.787073  0.000597  0.000165 -0.000066 -0.000442  0.023946  0.004480   
4  20.320969  0.001167  0.000239  0.001152 -0.000584  0.024612  0.004537   

         14        15  
0  0.228747  0.011801  
1  0.229715  0.011622  
2  0.230168  0.011512  
3  0.231609  0.011092  
4  0.232767  0.011054

In [15]:
Xc.shape

(11951, 16)

In [16]:
len(DATA_hs)

84

In [17]:
#Carregando arquivos
DATA_vs = []

for i in range(0,len(DATA_hs),3):
    VStacker(DATA_hs[i:i+3],(re.findall('\d*\D+',DATA_hs[i:i+3][0])[0]+str(i/3)))

In [18]:
len(DATA_vs)

28

In [19]:
Xc = np.load('DB_REPO/stacked/PK_Arthur Araujo Filho 42_hs.npy')
dfn = pd.DataFrame(data=Xc)
#dfn.head()

Xc[:35853].shape

(11951, 16)

for i in range(0,len(DATA_vs),1):
    print min(np.load('DB_REPO/stacked/'+DATA_vs[i:i+1][0]).shape[0],np.load('DB_REPO/stacked/'+DATA_vs[i:i+1][1]).shape[0],
              np.load('DB_REPO/stacked/'+DATA_vs[i:i+1][2]).shape[0],np.load('DB_REPO/stacked/'+DATA_vs[i:i+1][3]).shape[0],
              np.load('DB_REPO/stacked/'+DATA_vs[i:i+1][4]).shape[0], np.load('DB_REPO/stacked/'+DATA_vs[i:i+1][5]).shape[0])

i= 6
print min(np.load('DB_REPO/stacked/'+DATA_vs[i:i+6][0]).shape[0],np.load('DB_REPO/stacked/'+DATA_vs[i:i+6][1]).shape[0],
           np.load('DB_REPO/stacked/'+DATA_vs[i:i+6][2]).shape[0],np.load('DB_REPO/stacked/'+DATA_vs[i:i+6][3]).shape[0],
           np.load('DB_REPO/stacked/'+DATA_vs[i:i+6][4]).shape[0], np.load('DB_REPO/stacked/'+DATA_vs[i:i+6][5]).shape[0])

In [20]:
DATA_vs

['CT_Antonia Perpetuo Borges 0_vs.npy',
 'CT_Francisco das Chagas dos Santos 1_vs.npy',
 'CT_Gelson Amantos de Souza  2_vs.npy',
 'CT_Geraldo Dias de Oliveira 3_vs.npy',
 'CT_Iraci Rodrigues Soares 4_vs.npy',
 'CT_Joaquina de Sousa Freitas  5_vs.npy',
 'CT_Jose Martins de Sousa  6_vs.npy',
 'CT_Larcir Iria da Conceicao Santos 7_vs.npy',
 'CT_Magda Aparecida Ferreira 8_vs.npy',
 'CT_Manuel Saraiva da Rocha 9_vs.npy',
 'CT_Maria das Gracas Vieira 10_vs.npy',
 'CT_Maria Santana Estevam Moreira 11_vs.npy',
 'CT_Maria Veira Borges 12_vs.npy',
 'CT_Zenobia Barroso Feitosa 13_vs.npy',
 'PK_Arthur Araujo Filho 14_vs.npy',
 'PK_Boanerges Ramos Cunha 15_vs.npy',
 'PK_Carmino Pereira da Silva 16_vs.npy',
 'PK_Diniz Esmeraldo 17_vs.npy',
 'PK_Donia Aires 18_vs.npy',
 'PK_Fernando Francisco da Silva 19_vs.npy',
 'PK_Gastao Goncalves da Silva 20_vs.npy',
 'PK_Ismael Rodrigues do Carmo 21_vs.npy',
 'PK_Ivanildes Barbosa 22_vs.npy',
 'PK_Jose Fernandes Araujo 23_vs.npy',
 'PK_Jose Francisto Fontineli 

DATA_vs[78:78+6]

DATA_fs = []

for i in range(0,len(DATA_vs),6):
    HStackerF(DATA_vs[i:i+6],(re.findall('\d*\D+',DATA_vs[i:i+6][0])[0]+str(i/6)))

DATA_fs

In [21]:
def proccess_data(x,DATA,LABEL,NUM_SAMPLES = 50):
    
    Xc = np.load("DB_REPO/stacked/"+x)
    print '\nXc initial shape ', Xc.shape
    
    #Principal component analysis
    #pca = PCA(n_components=3)
    #pca.fit(Xc)
    #Xc = pca.transform(Xc)
    #print pca.explained_variance_ratio_
    #print ('Entropia: %3f' % (sum(pca.explained_variance_ratio_)))
    #print '\nXc PCA shape ', Xc.shape
    
    #Create temporal serie
    Xc = generate_envelope(Xc, NUM_SAMPLES)
    print 'Xc temporal-serie shape ', Xc.shape
    
    #Labeling the PKS level    
    C = (np.ones(len(Xc))*LABEL).reshape((len(Xc),1))
    Xc = np.hstack((Xc.reshape(Xc.shape),C.reshape((len(Xc),1)))) 
    print 'Xc labeled shape ', Xc.shape
    
    # Salving in file on the folder <DB_PKS/preproc/labeled>
    #np.save('./DB_REPO/labeled/'+x[:2]+'_tsl', Xc, allow_pickle=False)
    np.save('./DB_REPO/labeled/'+x[:-4]+'_'+str(y), Xc, allow_pickle=False)
    #print ''+x[:2]+'_tsl'
    print x[:-4]+'_'+str(y)
    #DATA.append(x[:2]+'_tsl.npy')
    DATA.append(x[:-4]+'_'+str(y))

---

#Validação
start = timeit.default_timer()

NUM_SAMPLES = 500
l = [1]
DATA = []

for x, y in zip(DATA_hs, l):
    proccess_data(x,DATA,NUM_SAMPLES,y)
    
stop = timeit.default_timer()
ti = stop - start 
print ('Tempo total para o envelopamento: %.2f min' % (ti/60))

DATA

Xc = np.load('DB_INTENC/preproc/labeled/p4_hs.npy_tsl.npy')
dfn = pd.DataFrame(data=Xc)
dfn.head()

---

In [22]:
#X = 'F'
#Y = 'F'
#Z = 'F'
#l = [1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1,
# X, X, X, X, X, X, 15, 15, 15, 15, 15, 15, Y, Y, Y, Y, Y, Y, 15, 15, 15, 15, 15, 15, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
# 3, 3, 3, 3, 3, 3, Z, Z, Z, Z, Z, Z, 1, 1, 1, 1, 1, 1, 4, 4, 4, 4, 4, 4, 1, 1, 1, 1, 1, 1]

#l = [1, 2, 15, 15, 2, 1, 15, 15, 1, 1, 3, 1, 4, 1]
l = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
#l = [2, 15, 4, 1, 0, 0, 15]
for x, y in zip(DATA_vs, l):
    print (x[:-4]+'_'+str(y))

CT_Antonia Perpetuo Borges 0_vs_0
CT_Francisco das Chagas dos Santos 1_vs_0
CT_Gelson Amantos de Souza  2_vs_0
CT_Geraldo Dias de Oliveira 3_vs_0
CT_Iraci Rodrigues Soares 4_vs_0
CT_Joaquina de Sousa Freitas  5_vs_0
CT_Jose Martins de Sousa  6_vs_0
CT_Larcir Iria da Conceicao Santos 7_vs_0
CT_Magda Aparecida Ferreira 8_vs_0
CT_Manuel Saraiva da Rocha 9_vs_0
CT_Maria das Gracas Vieira 10_vs_0
CT_Maria Santana Estevam Moreira 11_vs_0
CT_Maria Veira Borges 12_vs_0
CT_Zenobia Barroso Feitosa 13_vs_0
PK_Arthur Araujo Filho 14_vs_1
PK_Boanerges Ramos Cunha 15_vs_1
PK_Carmino Pereira da Silva 16_vs_1
PK_Diniz Esmeraldo 17_vs_1
PK_Donia Aires 18_vs_1
PK_Fernando Francisco da Silva 19_vs_1
PK_Gastao Goncalves da Silva 20_vs_1
PK_Ismael Rodrigues do Carmo 21_vs_1
PK_Ivanildes Barbosa 22_vs_1
PK_Jose Fernandes Araujo 23_vs_1
PK_Jose Francisto Fontineli  24_vs_1
PK_Marcilio Pontinho de Rezende 25_vs_1
PK_Osvaldir Borborema 26_vs_1
PK_Sergio Lamenza 27_vs_1


In [23]:
#Envelope size.

#Grupo total
start = timeit.default_timer()

DATA = []

for x, y in zip(DATA_vs, l):
    proccess_data(x,DATA,y)
    
stop = timeit.default_timer()
ti = stop - start 
print ('Tempo total para o envelopamento: %.2f min' % (ti/60))


Xc initial shape  (35853, 16)
Xc temporal-serie shape  (35804, 800)
Xc labeled shape  (35804, 801)
CT_Antonia Perpetuo Borges 0_vs_0

Xc initial shape  (35853, 16)
Xc temporal-serie shape  (35804, 800)
Xc labeled shape  (35804, 801)
CT_Francisco das Chagas dos Santos 1_vs_0

Xc initial shape  (35853, 16)
Xc temporal-serie shape  (35804, 800)
Xc labeled shape  (35804, 801)
CT_Gelson Amantos de Souza  2_vs_0

Xc initial shape  (35853, 16)
Xc temporal-serie shape  (35804, 800)
Xc labeled shape  (35804, 801)
CT_Geraldo Dias de Oliveira 3_vs_0

Xc initial shape  (35853, 16)
Xc temporal-serie shape  (35804, 800)
Xc labeled shape  (35804, 801)
CT_Iraci Rodrigues Soares 4_vs_0

Xc initial shape  (35853, 16)
Xc temporal-serie shape  (35804, 800)
Xc labeled shape  (35804, 801)
CT_Joaquina de Sousa Freitas  5_vs_0

Xc initial shape  (35853, 16)
Xc temporal-serie shape  (35804, 800)
Xc labeled shape  (35804, 801)
CT_Jose Martins de Sousa  6_vs_0

Xc initial shape  (39853, 16)
Xc temporal-serie sh

In [70]:
def pca_data(x,DATA,entrol,LABEL):
    
    Xc = np.load("DB_REPO/labeled/"+x+'.npy')
    Xc = Xc[:,:800]
    print '\nXc initial shape ', Xc.shape
    
    #Principal component analysis
    pca = PCA(n_components=100)
    pca.fit(Xc)
    Xc = pca.transform(Xc)
    print pca.explained_variance_ratio_
    print ('Entropia: %3f' % (sum(pca.explained_variance_ratio_)))
    print '\nXc PCA shape ', Xc.shape
    entro.append((sum(pca.explained_variance_ratio_)))
    
    #Labeling the PKS level    
    C = (np.ones(len(Xc))*LABEL).reshape((len(Xc),1))
    Xc = np.hstack((Xc.reshape(Xc.shape),C.reshape((len(Xc),1)))) 
    print 'Xc labeled shape ', Xc.shape
   
    # Salving in file on the folder <DB_PKS/preproc/labeled>
    #np.save('./DB_REPO/labeled/'+x[:2]+'_tsl', Xc, allow_pickle=False)
    np.save('./DB_REPO/pca/'+x[:-4]+'_'+str(y)+'pca', Xc, allow_pickle=False)
    #print ''+x[:2]+'_tsl'
    print x[:-4]+'_'+str(y)+'pca'
    #DATA.append(x[:2]+'_tsl.npy')
    DATA_pca.append(x[:-4]+'_'+str(y)+'pca')

In [71]:
#PCA
entro = []

start = timeit.default_timer()

DATA_pca = []

for x, y in zip(DATA, l):
    pca_data(x,DATA_pca, entro, y)
    
stop = timeit.default_timer()
ti = stop - start 
print ('Tempo total para o envelopamento: %.2f min' % (ti/60))


Xc initial shape  (35804, 800)
[ 0.45822796  0.044436    0.01976266  0.01408631  0.01165447  0.01153257
  0.0105412   0.01012086  0.00991677  0.0098396   0.00969482  0.00963181
  0.00946164  0.00940352  0.00928442  0.00914114  0.00905835  0.00902179
  0.00899015  0.00883158  0.00864941  0.0085273   0.00832677  0.00832305
  0.00820021  0.0081231   0.00804149  0.00792649  0.00790257  0.00778255
  0.0077594   0.0076359   0.00751268  0.00746591  0.00736659  0.00726381
  0.00725727  0.00717927  0.00715411  0.00706665  0.00705891  0.00702994
  0.00702038  0.00693928  0.00688328  0.00674569  0.00673271  0.00668355
  0.00666967  0.00639968  0.00636511  0.00461572  0.00426974  0.00284461
  0.00247874  0.00246539  0.00223655  0.00195974  0.00188605  0.00182798
  0.00178725  0.00162548  0.00157817  0.00153612  0.00142922  0.00142579
  0.00139151  0.00136602  0.00135911  0.00134673  0.00129551  0.00128712
  0.00122259  0.00118997  0.00118321  0.00118239  0.00115655  0.00115008
  0.00107561  0.001


Xc initial shape  (35804, 800)
[ 0.24417731  0.04182153  0.0268659   0.02291443  0.02241843  0.02185217
  0.01951798  0.01860307  0.01803033  0.01745478  0.01706615  0.01661332
  0.01653558  0.0162408   0.01600081  0.01532666  0.01508785  0.01492959
  0.01410134  0.01329252  0.0132879   0.01302561  0.01237437  0.01231172
  0.01200575  0.01195028  0.0118555   0.01178321  0.01167448  0.0116212
  0.01148533  0.01148041  0.01136208  0.01078569  0.01057929  0.01056304
  0.01050788  0.01037614  0.01035395  0.01022308  0.00984746  0.00935391
  0.00925298  0.00904671  0.00897239  0.00838921  0.0082229   0.00820336
  0.00804506  0.00753265  0.0064047   0.00223932  0.00178901  0.00163991
  0.00159838  0.00157635  0.00156976  0.00148513  0.00144656  0.00140582
  0.00134251  0.00132568  0.00129864  0.00128544  0.00127526  0.00126614
  0.00124807  0.00123202  0.00122145  0.00121348  0.00120859  0.00119203
  0.00118072  0.00117897  0.00117346  0.00117238  0.00113392  0.00113212
  0.00112619  0.0011


Xc initial shape  (35804, 800)
[ 0.06083736  0.02443321  0.02196851  0.02026603  0.02007599  0.01994895
  0.01988881  0.01979926  0.01977531  0.01951254  0.01940318  0.01915402
  0.01890817  0.0187237   0.01866007  0.01862072  0.0184813   0.01835505
  0.01823434  0.01822599  0.01821494  0.01816169  0.01814125  0.01813815
  0.01806633  0.0180282   0.01795753  0.0178319   0.01745621  0.01737031
  0.01733006  0.01731478  0.01725914  0.01717617  0.01715884  0.01707955
  0.01694889  0.01691027  0.01676765  0.01675657  0.01649821  0.01649174
  0.01619487  0.01614781  0.01611541  0.0160886   0.01578876  0.01572342
  0.01549733  0.01548647  0.00586824  0.00101023  0.00100577  0.00099377
  0.00098265  0.0009705   0.00095892  0.00095437  0.00095151  0.00094632
  0.0009452   0.00093306  0.00092266  0.00092206  0.00091894  0.00091619
  0.0009153   0.00091201  0.00091044  0.00090821  0.0009035   0.00090324
  0.00089937  0.00089613  0.00089366  0.00089344  0.000891    0.0008898
  0.00088092  0.0008


Xc initial shape  (35804, 800)
[ 0.18558391  0.01700938  0.01699024  0.01692783  0.01692572  0.0169182
  0.01617184  0.01614545  0.01607739  0.01606949  0.01599538  0.01599118
  0.01596775  0.01595143  0.01593584  0.01593518  0.01564914  0.01553208
  0.0155248   0.0155057   0.01549841  0.01548918  0.01546848  0.01540941
  0.01538879  0.01537336  0.0153173   0.0152666   0.01516828  0.01512802
  0.01504305  0.01502492  0.01493183  0.01486811  0.01483145  0.01482699
  0.01472869  0.01468884  0.01464756  0.01459563  0.01446436  0.01442727
  0.01408505  0.01401549  0.0140081   0.01376823  0.0137325   0.01327073
  0.01312551  0.01312076  0.00646476  0.00133233  0.00124109  0.00121037
  0.00120782  0.00120196  0.0011969   0.00118595  0.00118441  0.0011792
  0.00117777  0.00117722  0.00117588  0.00117522  0.00115969  0.00115499
  0.00114906  0.00113694  0.00113336  0.00113268  0.00112729  0.001127
  0.00112021  0.00111426  0.0011109   0.0011062   0.00110583  0.00110194
  0.00110056  0.0010993


Xc initial shape  (39804, 800)
[ 0.51152474  0.0292984   0.00817809  0.0077949   0.00647931  0.00621747
  0.00613757  0.00604335  0.00599183  0.00597135  0.00595615  0.00591916
  0.00590931  0.00586465  0.0058599   0.00584924  0.00579354  0.0057169
  0.0057057   0.00570538  0.00565303  0.00564101  0.00559707  0.00556326
  0.0055333   0.00550975  0.00550065  0.00550034  0.00549412  0.00547102
  0.00542688  0.0054204   0.00538795  0.00530946  0.0052928   0.00518646
  0.00518335  0.00515828  0.00514484  0.00513604  0.00508209  0.00506359
  0.00505615  0.00501869  0.00501348  0.00496827  0.00495982  0.00478142
  0.00477089  0.00474712  0.00473644  0.00468086  0.00465619  0.00437237
  0.00435922  0.00428594  0.0042829   0.00426511  0.00426076  0.00414927
  0.00411145  0.00410187  0.00409757  0.00408579  0.00407323  0.00406245
  0.00403562  0.00402837  0.00402561  0.00401319  0.00395263  0.00387685
  0.00385826  0.00385597  0.00379675  0.00377078  0.00373279  0.00371794
  0.00369865  0.0036

In [75]:
print ('Entropia média: %.5f' % np.mean(entro))
print ('Maior entropia: %.5f' % np.max(entro))
print ('Menor entropia: %.5f' % np.min(entro))

Entropia média: 0.99286
Maior entropia: 0.99963
Menor entropia: 0.94908


100 F -> 0.99285877528474364
120 F -> 0.99564845657083167
160 F -> 0.99895491841525164

In [26]:
1862.99/60/4*10

77.62458333333333

In [38]:
K = np.load("DB_REPO/labeled/"+DATA[1]+'.npy')

In [44]:
K[:,:800].shape

(35804, 800)

In [80]:
#Comentar caso não use janelamento
DATA = DATA_pca

#Não foram incluidas as amostras p4 e c3, serão utilizadas para validação!

In [81]:
DATA[2:-2]

['CT_Gelson Amantos de Souza  2__0pca',
 'CT_Geraldo Dias de Oliveira 3__0pca',
 'CT_Iraci Rodrigues Soares 4__0pca',
 'CT_Joaquina de Sousa Freitas  5__0pca',
 'CT_Jose Martins de Sousa  6__0pca',
 'CT_Larcir Iria da Conceicao Santos 7__0pca',
 'CT_Magda Aparecida Ferreira 8__0pca',
 'CT_Manuel Saraiva da Rocha 9__0pca',
 'CT_Maria das Gracas Vieira 10__0pca',
 'CT_Maria Santana Estevam Moreira 11__0pca',
 'CT_Maria Veira Borges 12__0pca',
 'CT_Zenobia Barroso Feitosa 13__0pca',
 'PK_Arthur Araujo Filho 14__1pca',
 'PK_Boanerges Ramos Cunha 15__1pca',
 'PK_Carmino Pereira da Silva 16__1pca',
 'PK_Diniz Esmeraldo 17__1pca',
 'PK_Donia Aires 18__1pca',
 'PK_Fernando Francisco da Silva 19__1pca',
 'PK_Gastao Goncalves da Silva 20__1pca',
 'PK_Ismael Rodrigues do Carmo 21__1pca',
 'PK_Ivanildes Barbosa 22__1pca',
 'PK_Jose Fernandes Araujo 23__1pca',
 'PK_Jose Francisto Fontineli  24__1pca',
 'PK_Marcilio Pontinho de Rezende 25__1pca']

In [82]:
len(DATA[2:-2])

24

In [83]:
DATA[-1]

'PK_Sergio Lamenza 27__1pca'

---

## Shape data

In [86]:
#Xc = np.load("DB_REPO/labeled/"+DATA[0]+'.npy')
Xc = np.load("DB_REPO/pca/"+DATA[0]+'.npy')
for i in DATA[2:-2]:
    #Xc = np.vstack((Xc,np.load('DB_REPO/labeled/'+i+'.npy')))
    Xc = np.vstack((Xc,np.load('DB_REPO/pca/'+i+'.npy')))

### Split tipo 1

#Dividindo os dados em sets de treino, test e validação
dfn = pd.DataFrame(data=Xc)
train, validate, test = np.split(dfn.sample(frac=1), [int(.6*len(dfn)), int(.8*len(dfn))])

np.save('./DB_INTENC/preproc/labeled/train', train, allow_pickle=False)
np.save('./DB_INTENC/preproc/labeled/validate', validate, allow_pickle=False)
np.save('./DB_INTENC/preproc/labeled/test', test, allow_pickle=False)

Xc_train = np.load("DB_INTENC/preproc/labeled/train.npy")

X_train = Xc_train[:,0:(Xc_train.shape[1]-1)]

yz = Xc_train[:,[(Xc_train.shape[1]-1)]]
y_train = np.array([])
for i in range(len(yz)):
    y_train = np.hstack((y_train,yz[i]))
    
lab_enc = preprocessing.LabelEncoder()
y_train = lab_enc.fit_transform(y_train)

X_train.shape, y_train.shape

#Xc_validate = np.load("DB_INTENC/preproc/labeled/validate.npy")
Xc_validate = Xc

X_validate = Xc_validate[:,0:(Xc_validate.shape[1]-1)]

yz = Xc_validate[:,[(Xc_validate.shape[1]-1)]]
y_validate = np.array([])
for i in range(len(yz)):
    y_validate = np.hstack((y_validate,yz[i]))
    
#lab_enc = preprocessing.LabelEncoder()
#y_validate = lab_enc.fit_transform(y_validate)

X_validate.shape, y_validate.shape

Xc_test= np.load("DB_INTENC/preproc/labeled/validate.npy")

X_test = Xc_test[:,0:(Xc_test.shape[1]-1)]

yz = Xc_test[:,[(Xc_test.shape[1]-1)]]
y_test = np.array([])
for i in range(len(yz)):
    y_test = np.hstack((y_test,yz[i]))
    
lab_enc = preprocessing.LabelEncoder()
y_test = lab_enc.fit_transform(y_test)

X_test.shape, y_test.shape

### Split tipo 2

In [87]:
print DATA[0]
print DATA[1]
print DATA[-1]
print DATA[-2]

CT_Antonia Perpetuo Borges 0__0pca
CT_Francisco das Chagas dos Santos 1__0pca
PK_Sergio Lamenza 27__1pca
PK_Osvaldir Borborema 26__1pca


In [88]:
for i in DATA[1:3]:
    print i

CT_Francisco das Chagas dos Santos 1__0pca
CT_Gelson Amantos de Souza  2__0pca


In [89]:
#Separando set de dados X, e set de labels y - Validação DATA[5] e DATA[6]
#Xc = np.load("DB_REPO/labeled/"+DATA[0]+'.npy')
Xc = np.load("DB_REPO/pca/"+DATA[0]+'.npy')

#for i in DATA[1:3]:
#    Xc = np.vstack((Xc,np.load('DB_REPO/labeled/'+i))) 

#Xc = np.vstack((Xc,np.load('DB_REPO/labeled/'+DATA[1]+'.npy')))
#Xc = np.vstack((Xc,np.load('DB_REPO/labeled/'+DATA[-1]+'.npy')))
#Xc = np.vstack((Xc,np.load('DB_REPO/labeled/'+DATA[-2]+'.npy')))

Xc = np.vstack((Xc,np.load('DB_REPO/pca/'+DATA[1]+'.npy')))
Xc = np.vstack((Xc,np.load('DB_REPO/pca/'+DATA[-1]+'.npy')))
Xc = np.vstack((Xc,np.load('DB_REPO/pca/'+DATA[-2]+'.npy')))

X_validate = Xc[:,0:(Xc.shape[1]-1)]

yz = Xc[:,[(Xc.shape[1]-1)]]
y_validate = np.array([])
for i in range(len(yz)):
    y_validate = np.hstack((y_validate,yz[i]))
    
X_validate.shape, y_validate.shape

((143216, 100), (143216,))

In [94]:
X_validate_std = sc.transform(X_validate)

X_validate_std.shape, y_validate.shape

((143216, 100), (143216,))

In [90]:
#Separando set de dados X, e set de labels y - Treino e Teste
#Xc = np.load("DB_REPO/labeled/"+DATA[3]+'.npy')
Xc = np.load("DB_REPO/pca/"+DATA[3]+'.npy')

#for i in DATA[1:-2]:
for i in DATA[4:-2]:
    #Xc = np.vstack((Xc,np.load('DB_REPO/labeled/'+i+'.npy'))) 
    Xc = np.vstack((Xc,np.load('DB_REPO/pca/'+i+'.npy'))) 
    
X = Xc[:,0:(Xc.shape[1]-1)]

yz = Xc[:,[(Xc.shape[1]-1)]]
y = np.array([])
for i in range(len(yz)):
    y = np.hstack((y,yz[i]))
    
X.shape, y.shape

((871492, 100), (871492,))

In [91]:
np.unique(y)

array([ 0.,  1.])

In [92]:
#Dividindo os dados em sets de treino e test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)

#X_test, X_validate, y_test, y_validate = train_test_split(X, y, test_size=0.5, random_state=1)

In [93]:
#Aplica normalizador P.S.: Verificar necessidade pois os dados já foram normalizados anteriormente
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
sc.fit(X_train)
X_train_std = sc.transform(X_train)
X_test_std = sc.transform(X_test)
X_validate_std = sc.transform(X_validate)

print X_train_std.shape, X_test_std.shape, X_validate_std.shape, y_train.shape, y_test.shape, y_validate.shape

(610044, 100) (261448, 100) (143216, 100) (610044,) (261448,) (143216,)


---

## Classifiers

### Random Forest

In [95]:
#Training and test
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier(n_estimators=100, n_jobs = 4)

print 'Training time'
%time rfc = rfc.fit(X_train_std, y_train)

print''

print 'Predict time'
%time y_pred = rfc.predict(X_test_std)

print''

from sklearn.metrics import accuracy_score
print ('ClassifyRF accuracy:---------->%.2f %%' % (accuracy_score(y_test, y_pred)*100))

Training time
CPU times: user 49min 25s, sys: 6.66 s, total: 49min 31s
Wall time: 13min 9s

Predict time
CPU times: user 23.9 s, sys: 244 ms, total: 24.1 s
Wall time: 6.53 s

ClassifyRF accuracy:---------->90.08 %


### Memória anterior
Training time
CPU times: user 4min 12s, sys: 1.13 s, total: 4min 13s
Wall time: 1min 20s

Predict time
CPU times: user 3.82 s, sys: 188 ms, total: 4 s
Wall time: 1.45 s

ClassifyRF accuracy:---------->98.35 %


#Salva a rede treinada para um arquivo
joblib.dump(rfc, 'DB_INTENC/preproc/labeled/rfc_split2.pkl') 

rfc = joblib.load('DB_INTENC/preproc/labeled/rfc.pkl')

In [96]:
#Validation
print 'Predict time'
%time y_pred = rfc.predict(X_validate)

print''

from sklearn.metrics import accuracy_score

result = ('ClassifyRF accuracy:---------->%.2f %%' % (accuracy_score(y_pred, y_validate)*100))

print result

Predict time
CPU times: user 2.98 s, sys: 52 ms, total: 3.03 s
Wall time: 1.17 s

ClassifyRF accuracy:---------->50.14 %


### Memória anterior
Predict time
CPU times: user 2.62 s, sys: 4 ms, total: 2.63 s
Wall time: 849 ms

ClassifyRF accuracy:---------->52.39 %


Predict time
CPU times: user 1.37 s, sys: 216 ms, total: 1.59 s
Wall time: 591 ms

ClassifyRF accuracy:---------->22.04 %


In [97]:
#Validation
print 'Predict time'
%time y_pred = rfc.predict(X_validate_std)

print''

from sklearn.metrics import accuracy_score

result = ('ClassifyRF accuracy:---------->%.2f %%' % (accuracy_score(y_pred, y_validate)*100))

print result

Predict time
CPU times: user 8.6 s, sys: 36 ms, total: 8.64 s
Wall time: 2.79 s

ClassifyRF accuracy:---------->61.79 %


### Memória anterior
Predict time
CPU times: user 9.48 s, sys: 12 ms, total: 9.49 s
Wall time: 2.63 s

ClassifyRF accuracy:---------->61.41 %

Predict time
CPU times: user 2.17 s, sys: 104 ms, total: 2.27 s
Wall time: 810 ms

ClassifyRF accuracy:---------->18.04 %

---

## Support Vector Machine

In [65]:
from sklearn import svm, datasets

print 'Processing time SVM'
%time svm = svm.SVC(kernel='rbf', gamma=0.2, C=6, decision_function_shape='ovr').fit(X_train_std, y_train)

print ''

from sklearn.metrics import accuracy_score

print 'Predict time'  
%time y_pred = svm.predict(X_test_std)
from sklearn.metrics import accuracy_score
print ('SVM-RBF accuracy:---------->%.2f %%' % (accuracy_score(y_test, y_pred)*100))

Processing time SVM
CPU times: user 1h 45min 48s, sys: 2.05 s, total: 1h 45min 50s
Wall time: 1h 45min 54s

Predict time
CPU times: user 11min 41s, sys: 48 ms, total: 11min 41s
Wall time: 11min 42s
SVM-RBF accuracy:---------->81.72 %


In [66]:
print 'Predict time'  
%time y_pred = svm.predict(X_validate_std)
from sklearn.metrics import accuracy_score
print ('SVM-RBF accuracy:---------->%.2f %%' % (accuracy_score(y_validate, y_pred)*100))

Predict time
CPU times: user 10min 56s, sys: 24 ms, total: 10min 56s
Wall time: 10min 56s
SVM-RBF accuracy:---------->6.63 %


## TESTS

In [79]:
from sklearn.ensemble import AdaBoostClassifier
print 'Processing time ADA'
%time ada = AdaBoostClassifier(n_estimators=600,learning_rate=0.3).fit(X_train_std, y_train)

print ''

print 'Predict time'  
%time y_pred = ada.predict(X_test_std)
print ('ADA accuracy:---------->%.2f %%' % (accuracy_score(y_test, y_pred)*100))

Processing time ADA
CPU times: user 6min 55s, sys: 52 ms, total: 6min 55s
Wall time: 6min 55s

Predict time
CPU times: user 15.4 s, sys: 12 ms, total: 15.5 s
Wall time: 15.5 s
ADA accuracy:---------->59.37 %


In [80]:
print 'Predict time'  
%time y_pred = ada.predict(X_validate_std)
print ('ADA accuracy:---------->%.2f %%' % (accuracy_score(y_validate, y_pred)*100))

Predict time
CPU times: user 13.9 s, sys: 0 ns, total: 13.9 s
Wall time: 13.9 s
ADA accuracy:---------->6.72 %


## 2

In [73]:
from sklearn.tree import DecisionTreeClassifier
print 'Processing time ADA'
%time ada = AdaBoostClassifier(DecisionTreeClassifier(max_depth=2),n_estimators=600,learning_rate=1).fit(X_train_std, y_train)

print ''

print 'Predict time'  
%time y_pred = ada.predict(X_test_std)
print ('ADA accuracy:---------->%.2f %%' % (accuracy_score(y_test, y_pred)*100))

Processing time ADA
CPU times: user 13min 4s, sys: 288 ms, total: 13min 4s
Wall time: 13min 5s

Predict time
CPU times: user 15 s, sys: 0 ns, total: 15 s
Wall time: 15 s
ADA accuracy:---------->39.56 %


In [74]:
print 'Predict time'  
%time y_pred = ada.predict(X_validate_std)
print ('ADA accuracy:---------->%.2f %%' % (accuracy_score(y_validate, y_pred)*100))

Predict time
CPU times: user 14.3 s, sys: 0 ns, total: 14.3 s
Wall time: 14.3 s
ADA accuracy:---------->10.59 %
